core/controllers/voice_artist.py

```diff
@@ -23,96 +23,115 @@
+        filename: str. The filename of the audio.
+        entity_type: str. The type of entity to which the audio belongs.
+        entity_id: str. The id of the entity to which the audio belongs.
+        user_id: str. The ID of the user saving the audio.
+
+    Raises:
+        Exception: If audio not supplied.
+        Exception: If the filename extension is unsupported.
+    """
+    allowed_formats = list(feconf.ACCEPTED_AUDIO_EXTENSIONS.keys())
+
+    if not raw_audio_file:
+        raise Exception('No audio supplied')
+    dot_index = filename.rfind('.')
+    extension = filename[dot_index + 1:].lower()
+
+    if dot_index == -1 or dot_index == 0:
+        raise Exception(
+            'No filename extension: it should have '
+            'one of the following extensions: %s' % allowed_formats)
+    if extension not in feconf.ACCEPTED_AUDIO_EXTENSIONS:
+        raise Exception(
+            'Invalid filename extension: it should have '
+            'one of the following extensions: %s' % allowed_formats)
+
+    tempbuffer = python_utils.string_io()
+    tempbuffer.write(raw_audio_file)
+    tempbuffer.seek(0)
+    try:
+        # For every accepted extension, use the mutagen-specific
+        # constructor for that type. This will catch mismatched audio
+        # types e.g. uploading a flac file with an MP3 extension.
+        if extension == 'mp3':
+            audio = mp3.MP3(tempbuffer)
+        else:
+            audio = mutagen.File(tempbuffer)
+    except mutagen.MutagenError:
+        # The calls to mp3.MP3() versus mutagen.File() seem to behave
+        # differently upon not being able to interpret the audio.
+        # mp3.MP3() raises a MutagenError whereas mutagen.File()
+        # seems to return None. It's not clear if this is always
+        # the case. Occasionally, mutagen.File() also seems to
+        # raise a MutagenError.
+        raise Exception('Audio not recognized as a %s file' % extension)
+    tempbuffer.close()
+
+    if audio is None:
+        raise Exception('Audio not recognized as a %s file' % extension)
+    if audio.info.length > feconf.MAX_AUDIO_FILE_LENGTH_SEC:
+        raise Exception(
+            'Audio files must be under %s seconds in length. The uploaded '
+            'file is %.2f seconds long.' % (
+                feconf.MAX_AUDIO_FILE_LENGTH_SEC, audio.info.length))
+    if len(set(audio.mime).intersection(
+            set(feconf.ACCEPTED_AUDIO_EXTENSIONS[extension]))) == 0:
+        raise Exception(
+            'Although the filename extension indicates the file '
+            'is a %s file, it was not recognized as one. '
+            'Found mime types: %s' % (extension, audio.mime))
+
+    mimetype = audio.mime[0]
+
+    # For a strange, unknown reason, the audio variable must be
+    # deleted before opening cloud storage. If not, cloud storage
+    # throws a very mysterious error that entails a mutagen
+    # object being recursively passed around in app engine.
+    del audio
```

Comment:
Actually I am confused. The code below doesn't reference the "audio" variable. So how does this help? Could you paste the error log in a comment here?

Reply:
Ah I see, thanks for that.

Reply:
Might be worth looking into the error though? Should we file it on Github?

Reply:
Should we file an error for this?

Reply:
Yeah, we can do but I don't have enough information on this as I'm not sure how I can find the error log as I cannot upload audio to GCS locally.

Reply:
Ok, then let's leave this as a comment.